### PR TITLE
Enable stateless XNNPACK convolutions.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36202 Enable torch_speed_benchmark to accept different memory formats.
* #35791 Enable stateless XNNPACK linear.
* **#35790 Enable stateless XNNPACK convolutions.**
* #36032 Mobile CPU allocator.

Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contiguous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821864](https://our.internmc.facebook.com/intern/diff/D20821864)

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -696,7 +705,15 @@ at::Tensor _convolution(
     }
 #endif
   } else if (input.device().type() == c10::DeviceType::CPU || input.device().type() == c10::DeviceType::CUDA) {
-    if (params.use_cpu_depthwise3x3_winograd(input, weight)) {
+    if (params.use_xnnpack(input, weight, bias)) {
+      // This is NOT the preferred usage pattern as it forces expensive and
+      // unnecessary weight prepacking per invocation.  Still if we end up here,
+      // which we shouldn't if we have correctly intercepted and replaced this
+      // usage through JIT, still this code path is more efficient for NHWC
+      // tensors than the alternatives.
```
Comment (dreiss) [inline aten/src/ATen/native/Convolution.cpp:713]: I'd suggest dropping this comment or making it much shorter.  This message needs to be sent to the model developers or PyTorch frontend developers, who won't be the ones reading this code.  "Using prepacked conv is preferred, but XNNPACK is still the fastest option for NHWC."

File: aten/src/ATen/native/Convolution.cpp [hunk 2]
```diff
@@ -696,7 +705,15 @@ at::Tensor _convolution(
     }
 #endif
   } else if (input.device().type() == c10::DeviceType::CPU || input.device().type() == c10::DeviceType::CUDA) {
```
Comment (dreiss) [inline aten/src/ATen/native/Convolution.cpp:707]: I'm confused.  We're doing a "use_xnnpack" check when the device type is CUDA?

**Review by dreiss — CHANGES_REQUESTED** [review]
This is awesome!  Just need to make sure we're not using this on CUDA.

**Comment by dreiss** [comment]
Oh, also need to see a test plan.  Are we sure this is a win on desktop/server mkldnn is available?

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:713]: Sounds good, will change.

File: aten/src/ATen/native/Convolution.cpp (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:707]: The check is if device type is CPU or CUDA, right?  use_cpu_depthwise3x3_winograd in this block is also CPU only.  I can break the check into two sections if you prefer.

**Comment by AshkanAliabadi** [comment]
@dreiss I should probably limit this to mobile builds only.  That would be the safer option.

**Comment by AshkanAliabadi** [comment]
@dreiss Alternatively we can enable XNNPACK for NHWC and NCHW both and remote dependency on NNPACK for binary size.  I will run a bunch of models to see how the performance compares.

File: aten/src/ATen/native/Convolution.cpp [hunk 3]
```diff
@@ -566,7 +580,7 @@ at::Tensor _convolution(
   auto k = weight.ndimension();
   c10::IntArrayRef weight_sizes = weight.sizes();
   int64_t dim = k - 2;
-  
+
```
Comment (kimishpatel) [inline aten/src/ATen/native/Convolution.cpp:583]: Empty space.

File: aten/src/ATen/native/Convolution.cpp [hunk 4]
```diff
@@ -583,7 +597,7 @@ at::Tensor _convolution(
 
   check_shape_forward(input, weight_sizes, bias, params);
 
-  if (input.size(0) == 0) {    
+  if (input.size(0) == 0) {
```
Comment (kimishpatel) [inline aten/src/ATen/native/Convolution.cpp:600]: Trailing whitespaces.

File: aten/src/ATen/native/Convolution.cpp [hunk 5]
```diff
@@ -696,7 +710,11 @@ at::Tensor _convolution(
     }
 #endif
   } else if (input.device().type() == c10::DeviceType::CPU || input.device().type() == c10::DeviceType::CUDA) {
-    if (params.use_cpu_depthwise3x3_winograd(input, weight)) {
+    if (params.use_xnnpack(input, weight, bias)) {
```
Comment (kimishpatel) [inline aten/src/ATen/native/Convolution.cpp:713]: It seems that for depthwise 3x3 they have ARM_NEON implementation of winowgrad. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L128. Do you know if xnnpack is still faster for depthwise compared to that implementation? Not sure if you saw this while doing experiments with mobilenet and resnet.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left a couple of comments.

File: aten/src/ATen/native/Convolution.cpp (hunk 5 above)
Comment (kimishpatel) [reply aten/src/ATen/native/Convolution.cpp:713]: Looks like you added that.

File: aten/src/ATen/native/Convolution.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:583]: Thanks will fix.

File: aten/src/ATen/native/Convolution.cpp (hunk 4 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:600]: Thanks will fix.

File: aten/src/ATen/native/Convolution.cpp (hunk 5 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:713]: Like everything performance I think the best solution would be not to make any assumptions and instead benchmark.  I will conduct a bunch of benchmarks and get back.

File: aten/src/ATen/native/Convolution.cpp (hunk 4 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:600]: OK, actually seems I removed (not added) a trailing space. :)  This is a setting on my editor to removing trailing spaces on save.

File: aten/src/ATen/native/Convolution.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:583]: Same here.  Removed trailing space on save.  Will disable this setting on the editor, but let's let this pass.  It feels wrong adding that trailing space back.

File: aten/src/ATen/native/Convolution.cpp [hunk 6]
```diff
@@ -239,6 +242,19 @@ auto ConvParams::use_nnpack(const at::Tensor& input) const -> bool {
   return false;
 }
 
+auto ConvParams::use_xnnpack(
+    const at::Tensor& input,
+    const at::Tensor& weight,
+    const at::Tensor& bias) const -> bool {
+// #if defined(C10_MOBILE)
```
Comment (dreiss) [inline aten/src/ATen/native/Convolution.cpp:249]: Shouldn't leave this comment in.

File: aten/src/ATen/native/Convolution.cpp (hunk 2 above)
Comment (dreiss) [reply aten/src/ATen/native/Convolution.cpp:714]: Yeah.  This code makes it look like use_xnnpack or use_cpu_depthwise3x3_winograd might return true on a CUDA input.  I think a clearer option would be
```
} else if (is_cpu && use_xnn) {
  ...
} else if (is_cpu && use_winograd) {
  ...
} else if (is_cpu || is_cuda) {
  ...
}
```

Or even drop the `DeviceType::CPU` checks and move them into `use_xnnpack` and `use_winograd`

**Review by dreiss — CHANGES_REQUESTED** [review]
Can you include a test plan in the commit message?

File: aten/src/ATen/native/Convolution.cpp (hunk 6 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:249]: No just trying to get CI coverage on non-mobile builds to fix potential issues.

File: aten/src/ATen/native/Convolution.cpp (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp:714]: Sounds good, will change.

File: aten/src/ATen/native/xnnpack/Convolution.cpp [hunk 7]
```diff
@@ -154,7 +155,9 @@ ContextConv2D create(
       weight_nhwc.size(Layout::Filter::input) * groups,               // input_pixel_stride
       weight_nhwc.size(Layout::Filter::output),                       // output_pixel_stride
       weight_nhwc.data_ptr<float>(),                                  // kernel
-      (bias && bias->defined()) ? bias->data_ptr<float>() : nullptr,  // bias
+      (bias && bias->defined())
+          ? bias->contiguous().data_ptr<float>()
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Convolution.cpp:159]: Wait, this doesn't seem safe.  If bias is non-contiguous, won't this contiguous call allocate a new tensor, which gets freed right after the end of this line?

File: aten/src/ATen/native/Convolution.cpp [hunk 8]
```diff
@@ -695,11 +710,30 @@ at::Tensor _convolution(
...
+    output = xnnpack::convolution2d(
+        input,
+        weight,
+        bias,
+        params.padding,
+        params.stride,
+        params.dilation,
+        params.groups);
+  } else if (params.use_cpu_depthwise3x3_winograd(input, weight)) {
+    output = convolution_depthwise3x3_winograd_stub(
```
Comment (dreiss) [inline aten/src/ATen/native/Convolution.cpp:725]: Can fix this in a separate diff, but it looks like this doesn't work if bias is on GPU, and maybe also doesn't handle non-contiguous input,weight,bias.

**Review by dreiss — CHANGES_REQUESTED** [review]
Looks good except for what I think is a use-after-free.

File: aten/src/ATen/native/xnnpack/Convolution.cpp (hunk 7 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Convolution.cpp:159]: My reading of the standard is that this usage is safe.  From [Temporary Object Lifetime](https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime):

> All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation. This is true even if that evaluation ends in throwing an exception.

Where a full-expression according to the [C++ standard [draft]](https://timsong-cpp.github.io/cppwp/n3337/intro.execution#10) is defined as:

> A full-expression is an expression that is not a subexpression of another expression.

Stack Overflow question: https://stackoverflow.com/a/6059134
Example: http://cpp.sh/9tn6d

I can pull this out and convert to below as well:

```
if (bias && bias->defined()) {
  bias = bias->contiguous();
}
```

Comment (dreiss) [reply aten/src/ATen/native/xnnpack/Convolution.cpp:159]: Wait, does xnn_create_convolution2d_nhwc_f32 copy the bias?  If so, then you're definitely right.  For some reason I was assuming it retained a pointer to the original bias data.

**Review by dreiss — APPROVED** [review]

**Pull request closed without merging**